Bool tensor. Part 0: Boolean storage implementation

izdeby opened the pull request:
This is the first commit from a series of planned changes in order to add boolean tensors to PyTorch. The whole plan looks like this: 

0. Storage Implementation (this change)
1. Tensor Creation.
2. Tensor Conversions.
3. Tensor Indexing.
4. Tensor Operations.
5. Back compatibility related changes.

This feature was requested by the community: 
https://github.com/pytorch/pytorch/issues/4764
https://github.com/pytorch/pytorch/issues/4219
https://github.com/pytorch/pytorch/issues/4288

**Change**: 
Added boolean type to the Storage class for CPU and CUDA backends. 

**Tested via**: 
1. unit tests
2. running this: 
-> import torch
-> torch.BoolStorage
<class 'torch.BoolStorage'>
-> torch.cuda.BoolStorage
<class 'torch.cuda.BoolStorage'>

fmassa commented:
Question: do we want to consider the bool tensor as holding 1bit data? E.g., by having `uint8_t` and packing it as 8 element tensor?

albanD commented:
From a quick look, what is the reason why `THCGenerateBoolType.h` did not make it into `THGenerateAllTypes.h` ? Are there cases where you want to generate all other types but not bool?

izdeby commented:
> From a quick look, what is the reason why `THCGenerateBoolType.h` did not make it into `THGenerateAllTypes.h` ? Are there cases where you want to generate all other types but not bool?

This was done to minimize the amount of needed changes. If THCGenerateBoolType.h/THGenerateBoolType.h were included in THCGenerateAllTypes.h/THGenerateAllType.h, this change would me significantly bigger and harder to review. 
As an example, take a look at aten/src/TH/THTensor.h. If i had included bool type into *GenerateAllType, i would have to implement all Math, Random and Convolution functionality. This will be done, but later.

SsnL commented on torch/csrc/byte_order.cpp:
```diff
@@ -162,6 +162,11 @@ void THP_decodeDoubleBuffer(double* dst, const uint8_t* src, THPByteOrder order,
 
+void THP_decodeBoolBuffer(bool* dst, const uint8_t* src, THPByteOrder order, size_t len)
+{
+  dst = (bool*)src;
```
this looks different from other similar functions in this file

izdeby replied:
Fixed and added a test

izdeby commented:
> Question: do we want to consider the bool tensor as holding 1bit data? E.g., by having `uint8_t` and packing it as 8 element tensor?

@fmassa , it's 1 byte, which is the same as numpy.

apaszke commented:
I understand that it might be simpler to use `uint8_t` to represent a single element for now, but if we wanted to change the memory representation of those tensors in the future it would be a breaking change (people might have kernels), so it would be good to get it right sooner than later.

albanD commented:
@apaszke but that would mean that we could not do `from_numpy` on numpy.bool into this type without copy? Which is what at least one of the issue is about no?

fmassa commented:
A couple more thoughts:
- if we actually want to have binary masks to be part of autograd (e.g., in order to solve problems like https://github.com/pytorch/pytorch/issues/12986) or for the mask in batched tensors, it might be useful considering a 1bit representation for memory savings.
- having some kind of support for numeric representations with < 1 byte could be interesting, as it could potentially be a way of implementing low-precision numbers by packing 2/4/8 elements into a single byte

But I do agree that supporting 1bit (or 2bit? 4 bit?, 3 bit?) elements would be a much bigger endeavor. In particular, it would also mean changing all the support we currently have in TensorIterator and alike, which would not be easy.

gchanan commented:
I buy @albanD's argument; there are use cases for 1 byte bools (even if we could magically also have 1-bit bools), and given our numpy affinity, it seems like the right choice to call 1-byte bools `torch.bool`.

There's nothing precluding us from coming in later and supporting 1-bit bools via a packed representation or something (e.g. 1-bit quantization is a thing) that wouldn't break backwards compatibility.

apaszke commented:
@albanD yeah in that case `from_numpy` won't work on those tensors (it doesn't work today, and I don't think it's a terrible loss). I really think that the memory savings are quite nice, and otherwise we're just duplicating `torch.ByteTensor` with no clear distinction

izdeby commented on aten/src/TH/THTensor.cpp:
```diff
@@ -6,6 +6,9 @@
 #include <TH/generic/THTensor.cpp>
 #include <TH/THGenerateHalfType.h>
 
+#include <TH/generic/THTensor.cpp>
```
I need these THTensor/THCTensor changes in order to build THCStorageCopy.

izdeby commented on c10/core/ScalarType.h:
```diff
@@ -24,7 +24,8 @@ _(float,Float,d)   /* 6 */ \
 _(std::complex<float>,ComplexFloat,z)   /* 9 */ \
-_(std::complex<double>,ComplexDouble,z) /* 10 */
+_(std::complex<double>,ComplexDouble,z) /* 10 */ \
+_(bool,Bool,i) /* 11 */
```
Updated according to the NumPy's behavior

izdeby commented on aten/src/THC/THCTensorCopy.cu:
```diff
@@ -8,7 +8,7 @@
 struct CopyOp {
   __device__ __forceinline__ void operator()(TypeDst* dst, TypeSrc* src) {
-#if __CUDA_ARCH__ >= 350
+#if __CUDA_ARCH__ >= 350 && defined(THE_REAL_IS_BOOL)
```
@gchanan, yes. Without this, im getting this error: 
```
src/THC/THCTensorCopy.cu(12): error: no instance of overloaded function "__ldg" matches the argument list
            argument types are: (__nv_bool *)
...
```

albanD commented:
If the goal is just to be compatible with numpy bool and allow for better advanced indexing, do actually need to generate code a completely new scalar type (which looks like this PR is doing)? Can it be a small wrapper around ByteTensor with custom functions for numpy interface and so that it can be used by indexing?

gchanan commented:
@albanD I don't really see how you'd make that work nicely at the C++ level.

albanD commented:
I'm not familiar enough with the way aten works so that was just a suggestion in case it was possible.
Couldn't an implementation inherit from another one?

gchanan commented:
@albanD well C++ is a little more complicated :).  i.e. you'd have to make sure everything you need is virtual, which would slow down the main line path.  It seems very likely to be a leaky abstraction.

albanD commented:
Ok, good to know !
Larger binaries it is then !

gchanan commented:
@albanD I think the right thing to do is to only define minimal mathematical operators so the code size difference shouldn't be too big; people can `.to(torch.uint8)` if they really want to do math.

izdeby commented on torch/csrc/generic/StorageMethods.cpp:
```diff
@@ -169,6 +169,8 @@ static PyObject * THPStorage_(fromBuffer)(PyObject *_unused, PyObject *args, PyO
 #elif defined(TH_REAL_IS_HALF)
   THP_decodeHalfBuffer(THWStorage_(data)(storage), src + offset, byte_order, count);
+#elif defined(TH_REAL_IS_BOOL)
+  THP_decodeBoolBuffer(THWStorage_(data)(storage), src + offset, byte_order, count);
```
Copying the buffer as discussed offline. Added test cases.

gchanan commented:
looks like you may have a clang_tidy issue.

gchanan commented on aten/src/TH/THGenerateBoolType.h:
```diff
@@ -0,0 +1,22 @@
+#endif
+
+#define scalar_t bool
+#define accreal bool
```
maybe, but I think that implies we shouldn't define `accreal` here (so it's obvious when we try to do math), not to define it as bool.

gchanan commented on torch/csrc/byte_order.cpp:
```diff
@@ -1,6 +1,7 @@
 #include <torch/csrc/byte_order.h>
 
 #include <cstring>
+#include <iostream>
```
I don't think you need the changes in this file.

gchanan commented on torch/csrc/utils.h:
```diff
@@ -81,6 +81,12 @@
 #define THPHalfUtils_unpackAccreal(object)    (double)THPUtils_unpackReal_FLOAT(object)
 #define THPHalfUtils_newAccreal(value)        THPUtils_newReal_FLOAT(value)
 
+#define THPBoolUtils_checkReal(object)        THPUtils_checkReal_INT(object)
```
I'm not sure these are right, because most of these call e.g. PyInt_Check / PyFloat_Real but you probably want to use PyBool_Check.

izdeby replied:
These need to stay INT as the code that is referencing these macros is expecting certain behavior. As an example, checkReal_* is being used in several places including StorageMethods.cpp in fill_() method. If we use PyBool_Check, fill_ method wont work as expected. There is a test that verifies that (it will fail in case of PyBool_Check) - test_print test in test_torch.py. Same holds for the unpackReal_* macro.

gchanan replied:
I don't think I agree.  No existing code references this because it is new code.  The fill_() example is trying to fill a BoolStorage with an int; it's debatable what the correct behavior is here, but we can look at existing storage examples:

1) Float accepts integers or doubles
2) Int accepts only integers (not float)

Given the restrictiveness of ints, I'd make bools restrictive too.  Plus it's always easier to make something less restrictive than more restrictive (because of backwards compatibility concerns).

Note that the Tensor behavior is different than Storage, but we can just focus on Storage for now.

gchanan commented on torch/csrc/utils.h:
```diff
@@ -81,6 +81,12 @@
 #define THPHalfUtils_newAccreal(value)        THPUtils_newReal_FLOAT(value)
 
+#define THPBoolUtils_checkReal(object)        THPUtils_checkReal_INT(object)
+#define THPBoolUtils_unpackReal(object)       (int64_t)THPUtils_unpackReal_INT(object)
```
you probably want a bool cast.

izdeby commented on aten/src/ATen/gen.py:
```diff
@@ -178,6 +178,7 @@ def check_all_files_written(self):
 
 # scalar_name, c_type, accreal, th_scalar_type, is_floating_type
 scalar_types = [
+    ('Bool', 'bool', 'Bool', 'bool', False),
```
We need this as gen.py generates core/Type.h where CPUBool, SparseCPUBool, CUDABool and SparseCUDABool are defined. 
Changing to be ('Bool', 'uint8_t', 'Long', 'uint8_t', False),

gchanan replied:
What about just not defining it to be anything valid?  I'm pretty certain this won't actually be used in codegen, just make it something like BoolAccrealNotDefined.

izdeby commented on c10/core/ScalarType.h:
```diff
@@ -185,6 +186,7 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
   constexpr auto f2 = ScalarType::Half;
   constexpr auto f4 = ScalarType::Float;
   constexpr auto f8 = ScalarType::Double;
+  constexpr auto boolean = ScalarType::Bool;
```
'bool' is a reserved C++ keyword

gchanan replied:
whoops, was thinking C :).  How about "b1" so it doesn't mess with the spacing in the rest of the table?

gchanan commented on aten/src/TH/THGenerateBoolType.h:
```diff
@@ -0,0 +1,22 @@
+
+// TODO: define accreal type once the correct value is known.
+#define scalar_t bool
+#define ureal uint8_t
```
shouldn't this be bool? bools are already unsigned.

gchanan commented on aten/src/THC/THCGenerateBoolType.h:
```diff
@@ -0,0 +1,21 @@
+
+// TODO: define accreal type once the correct value is known.
+#define scalar_t bool
+#define ureal uint8_t
```
shouldn't this be bool?  bools are already unsigned.

izdeby commented on torch/csrc/generic/StorageMethods.cpp:
```diff
@@ -96,7 +96,7 @@ static PyObject * THPStorage_(fromBuffer)(PyObject *_unused, PyObject *args, PyO
   const char* byte_order_str = nullptr;
   Py_ssize_t count = -1, offset = 0;
-  Py_buffer buffer;
+  Py_buffer buffer = {};
```
Because i've touched this file, i was getting a clang-tidy error on this line. This is my attempt to fix it but i need fully green build to verify that it works.

izdeby commented on test/test_torch.py:
```diff
@@ -8850,6 +8853,21 @@ def test_from_buffer(self):
         self.assertEqual(floats[0], 2.25)
 
+        f = bytearray([0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x10, 0x40])
+        bools = torch.BoolStorage.from_buffer(f, 'big')
```
You are right, we dont need to pass it but as from_buffer is auto generated from a template, it expects two values to be passed. Its even smart enough to give an error in case some unexpected value is being passed instead if 'big'. Error: "ValueError: invalid byte_order '' (expected 'big', 'little', or 'native')'

izdeby commented on torch/storage.py:
```diff
@@ -83,6 +83,10 @@ def byte(self):
         return self.type(type(self).__module__ + '.ByteStorage')
 
+    def bool(self):
+        """Casts this storage to bool type"""
```
Not yet. I need to add BoolTensor support first. 
The way i want to test will look something like this: 

1: intTensor = torch.tensor([1,2,3,4], dtype=torch.int)
2: boolTensor = intTensor.bool()
3: // check if boolTensor is what we expect it is. 

But as of right now, im getting this error on line 2: "AttributeError: 'Tensor' object has no attribute 'bool'" 

Let me know if you want me to fix it in this PR.

gchanan replied:
But this is on Storage.  E.g.:
```
In [12]: torch.IntStorage(5).bool()
Out[12]:
 True
 True
 True
 True
 True
[torch.BoolStorage of size 5]
```

izdeby replied:
added a test

izdeby commented on test/test_torch.py:
```diff
@@ -8850,6 +8853,21 @@ def test_from_buffer(self):
+        self.assertEqual(bools.size(), 19)
+
+        f = bytearray(b'\0x4A')
+        bools = torch.BoolStorage.from_buffer(f, 'big')
```
So the problem here is with the '\' character. 
Here is what is going on:  
>>> f = bytearray(b'\0x4A')
>>> f
bytearray(b'\x00x4A')

>>> f = bytearray(b'0x4A')
>>> f
bytearray(b'0x4A')
-----

I checked against NumPy and the behavior is identical.

gchanan commented on aten/src/THC/THCTensorCopy.cu:
```diff
@@ -8,12 +8,23 @@
+};
+
+template <>
+struct CopyOp <bool, bool> {
```
I wouldn't fully specialize this unless you rewrite the above to only have one template parameter.  You can just partially specialize this.

gchanan replied:
The reason being the above won't actually work if someone was to instantiate <non_bool, bool>.

gchanan commented on test/test_torch.py:
```diff
@@ -8868,6 +8868,52 @@ def test_from_buffer(self):
         self.assertEqual(bools.tolist(), [False, True, True, True])
 
+    def test_storage_casts(self):
+        storage = torch.IntStorage([-1, 0, 1, 2, 3, 4])
```
this test is a bit strange.  I'd expect you'd want to check conversion to/from bool or do some exhaustive checking, but this is a bit of a strange combination (because you rewrite storage each time).  I'd just check you can convert to/from bool.

izdeby replied:
I just wanted to cover storage.<T> for all types. There is a part where i convert values to bool and back and show that the original value was changed. Or did you mean something else?

gchanan replied:
My point is this isn't a great way to test, because once you convert to bool, you can't actually check that the other types can handle values that aren't 0 or 1.  So it looks like you are exhaustively testing storages, but you aren't.  Thus the suggestion to either test storages exhaustively or just test conversion to/from bool.

gchanan commented on torch/csrc/utils.h:
```diff
@@ -41,6 +41,14 @@
     PyInt_Check(object) ? PyInt_AsLong(object) :                               \
     (throw std::runtime_error("Could not parse real"), 0))
+
+#define THPUtils_unpackReal_BOOL(object)                                       \
```
you don't need to define a version for py2 vs py3, bools are the same across versions.  Just put this below, outside of the if/endif.

gchanan commented on torch/csrc/generic/StorageMethods.cpp:
```diff
@@ -161,6 +161,8 @@ static PyObject * THPStorage_(fromBuffer)(PyObject *_unused, PyObject *args, PyO
 #if defined(TH_REAL_IS_BYTE) || defined(TH_REAL_IS_CHAR)
   memcpy(THWStorage_(data)(storage), src + offset, count);
 #elif defined(TH_REAL_IS_BOOL)
+  // Because of ASAN checks, we have to manually move bytes instead of
```
you aren't moving bytes though, you are changing the the values.  How about include the ASAN check that tripped this?

gchanan reviewed (approved):
changes look good, nice work!

I think the test failures will go away if you merge in master.

izdeby committed:
Bool tensor. Part 0: Boolean storage implementation (#16810)

Summary:
This is the first commit from a series of planned changes in order to add boolean tensors to PyTorch. The whole plan looks like this:

0. Storage Implementation (this change)
1. Tensor Creation.
2. Tensor Conversions.
3. Tensor Indexing.
4. Tensor Operations.
5. Back compatibility related changes.

This feature was requested by the community:
https://github.com/pytorch/pytorch/issues/4764
https://github.com/pytorch/pytorch/issues/4219
https://github.com/pytorch/pytorch/issues/4288

**Change**:
Added boolean type to the Storage class for CPU and CUDA backends.

**Tested via**:
1. unit tests
2. running this:
-> import torch
-> torch.BoolStorage
<class 'torch.BoolStorage'>
-> torch.cuda.BoolStorage
<class 'torch.cuda.BoolStorage'>
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16810

Differential Revision: D14087246

fbshipit-source-id: 12f22b897c33defddad1c967e4cf5ae764b85d13
